bpf: Report original source IP in TRACE_TO_LXC

tgraf opened the pull request:
In order to correlate flow events on multiple nodes, a possible service IP
translation must be known. The TRACE_TO_LXC event is situated in a position to
report this as reverse translation has been performed just before. Extend
the trace_notify structure to include the original source IP if available.

Example:
```
-> endpoint 2614 flow 0x5b9d9e5a identity 56995->31468 state new ifindex lxcc3be4eeac8fe orig-ip 10.16.182.13: 10.16.182.13:33020 -> 10.16.228.151:80 tcp SYN
-> endpoint 3745 flow 0xa4419b6f identity 31468->56995 state reply ifindex lxc155152faaf3f orig-ip 10.16.228.151: 172.20.0.187:80 -> 10.16.182.13:33020 tcp ACK, RST
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9321)
<!-- Reviewable:end -->

joestringer commented on pkg/monitor/datapath_trace.go:
```diff
@@ -17,13 +17,15 @@ package monitor
 const (
 	// TraceNotifyLen is the amount of packet data provided in a trace notification
-	TraceNotifyLen = 32
+	TraceNotifyLen = 48
```
Do we enforce recreation of the perf event channel on startup based upon this? Otherwise is there a guarantee upon upgrade that the new agent-side monitor won't get confused by the existing datapath pumping old-style trace notifications into the perf events channel?

aanm replied:
Good point :-/

tgraf replied:
It must be. The memory is owned by the user space process so it must get re-created when the agent restarts, no?

tgraf replied:
I've tested this. I did not observe any problems restarting between master and this PR branch.

aanm commented on pkg/monitor/datapath_trace.go:
```diff
@@ -108,16 +111,25 @@ func (n *TraceNotify) traceSummary() string {
+// OriginalIP returns the original source IP if reverse NAT was performed on
+// the flow
+func (n *TraceNotify) OriginalIP() net.IP {
+	if n.Flags != 0 {
```
Please check if `if n.Flags & ipv6Flag != 0` instead to avoid long debugging sessions in the future if we add one more flag.

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -1023,11 +1023,11 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, _
 		*proxy_port = verdict;
 		// Trace the packet before its forwarded to proxy
-		send_trace_notify(skb, TRACE_TO_PROXY, src_label, SECLABEL,
+		send_trace_notify4(skb, TRACE_TO_PROXY, src_label, SECLABEL, orig_sip,
```
I think you miss doing the same for `ipv6_policy`

tgraf replied:
Yes, I wanted to fully validate the concept first and make sure that the CI remains stable.

tgraf committed:
bpf: Report original source IP in TRACE_TO_LXC

In order to correlate flow events on multiple nodes, a possible service IP
translation must be known. The TRACE_TO_LXC event is situated in a position to
report this as reverse translation has been performed just before. Extend
the trace_notify structure to include the original source IP if available.

Example:
```
-> endpoint 2614 flow 0x5b9d9e5a identity 56995->31468 state new ifindex lxcc3be4eeac8fe orig-ip 10.16.182.13: 10.16.182.13:33020 -> 10.16.228.151:80 tcp SYN
-> endpoint 3745 flow 0xa4419b6f identity 31468->56995 state reply ifindex lxc155152faaf3f orig-ip 10.16.228.151: 172.20.0.187:80 -> 10.16.182.13:33020 tcp ACK, RST
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/monitor/datapath_trace.go:
```diff
@@ -108,16 +117,25 @@ func (n *TraceNotify) traceSummary() string {
 
+// OriginalIP returns the original source IP if reverse NAT was performed on
+// the flow
+func (n *TraceNotify) OriginalIP() net.IP {
```
```suggestion
func (n *TraceNotify) OriginalSourceIP() net.IP {
```

Just in case we want to encode the original destination at some point to understand that outbound monitor messages were in fact destined for a service.

(Alternatively we could just call it `ServiceIP` and make sure that the datapath populates it with the service IP, then repurpose the src/dst as pod IPs)

joestringer replied:
I think this is fine for now.

joestringer committed:
bpf: Refactor common capture header fields

We'll extend these in a subsequent commit.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Repurpose upper bits of cap_len for version

The cap_len in practice was previously always TRACE_PAYLOAD_LEN if a
notification message is emitted, which by default is 128 and was not
known to exceed 65535 bytes. By repurposing the upper 16 bits of this
field, we can now inject a version of the message which allows clients
to support extensions of the trace notify header to accommodate new
metadata.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
monitor: Extend trace notification version

Extend pkg/monitor to provide a generic DecodeTraceNotify() function
which receives a TraceNotify type (which supports both v0 and v1), and
reads into it the data from the monitor regardless of whether the
incoming messages are v0 or v1.

The version 0 is reused for existing messages to avoid breaking the
format for those messages; version 1 includes the original source
address.

Note that this may cause unnecessary dissection when using an older
monitor client with a newer version of Cilium, but in practice the
library that Cilium provides handled this in a reasonable manner in past
versions, and also it's more likely for users to run newer clients with
older Cilium, so this problem is largely ignored for this commit.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/monitor/datapath_trace.go:
```diff
@@ -77,9 +94,39 @@ func connState(reason uint8) string {
 func (n *TraceNotify) encryptReason() string {
 	if (n.Reason & TraceReasonEncryptMask) != 0 {
-		return fmt.Sprintf("encrypted ")
+		return fmt.Sprintf(" encrypted ")
```
Should the space be added somewhere else?

joestringer replied:
Oh, I had another minor cosmetic fix that was supposed to go with this, but I dropped the other part from this PR and accidentally left this in.

jrfastab commented on bpf/lib/common.h:
```diff
@@ -256,11 +256,16 @@ enum {
 #define NOTIFY_COMMON_HDR \
-	__u8		type; \
-	__u8		subtype; \
-	__u16		source; \
+	__u8		type;		\
+	__u8		subtype;	\
+	__u16		source;		\
 	__u32		hash;
```
Presumably this is backwards compatible? I didn't run through code to check but add a note in the commit msg might help if you end up doing another spin of this set.

jrfastab commented on pkg/monitor/datapath_trace.go:
```diff
@@ -77,6 +94,36 @@ func connState(reason uint8) string {
+		return err
+	}
+
+	switch msgHdrLen {
```
Got a bit confused here. So this basically uses the msgHdrLen as a version type for the decoder? Any reason why not to call it `msgHdrVersion` and just enum that field with the supported types? Otherwise in the future if we ever add another message type that happens to have the same size as V2Len it will be thrown off.

tgraf commented on bpf/lib/common.h:
```diff
@@ -264,7 +264,11 @@ enum {
 	__u32		len_orig;	/* Length of original packet */	\
-	__u32		len_cap;	/* Length of captured bytes */
+	__u16		len_cap;	/* Length of captured bytes */	\
+	__u16		version;	/* Capture header version */
```
I was originally considering to add a `hdr_len` field but a version is fine as well.

joestringer replied:
See discussion with John here, I think it's actually a bit cleaner this way.

https://github.com/cilium/cilium/pull/9321#discussion_r335526797
